aws-storage-s3/src/androidTest/java/com/amplifyframework/storage/s3/AWSS3StorageUploadTest.java

```diff
@@ -0,0 +1,238 @@
+        String smallFileKey = getS3Key(DEFAULT_ACCESS_LEVEL, SMALL_FILE_NAME);
+        cleanUpS3Object(largeFileKey);
+        cleanUpS3Object(smallFileKey);
+    }
+
+    /**
+     * Sets up the options to use for transfer.
+     */
+    @Before
+    public void setUp() {
+        // Always interact with PUBLIC access for consistency
+        options = StorageUploadFileOptions.builder()
+                .accessLevel(DEFAULT_ACCESS_LEVEL)
+                .build();
+
+        // Create a set to remember all the subscriptions
+        subscriptions = new HashSet<>();
+    }
+
+    /**
+     * Unsubscribe from everything after each test.
+     */
+    @After
+    public void unsubscribe() {
+        // Unsubscribe from everything
+        for (SubscriptionToken token : subscriptions) {
+            Amplify.Hub.unsubscribe(token);
+        }
+    }
+
+    /**
+     * Tests that small file (single-part) uploads successfully.
+     *
+     * @throws Exception if upload fails
+     */
+    @Test
+    public void testUploadSmallFile() throws Exception {
+        latchedUploadAndConfirm(smallFile, DEFAULT_ACCESS_LEVEL, getIdentityId());
+    }
+
+    /**
+     * Tests that large file (multi-part) uploads successfully.
+     *
+     * @throws Exception if upload fails
+     */
+    @Test
+    public void testUploadLargeFile() throws Exception {
+        latchedUploadAndConfirm(largeFile, DEFAULT_ACCESS_LEVEL, getIdentityId());
+    }
+
+    /**
+     * Tests that file upload operation can be canceled while the
+     * transfer hasn't completed yet.
+     *
+     * @throws Exception if upload is not canceled successfully
+     *         before timeout
+     */
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testUploadFileIsCancelable() throws Exception {
+        final CountDownLatch canceled = new CountDownLatch(1);
+
+        // Begin uploading large file
+        StorageUploadFileOperation<?> op = Amplify.Storage.uploadFile(
+            largeFile.getName(),
+            largeFile.getAbsolutePath(),
+            options,
+            onResult -> fail("Upload finished before being successfully cancelled."),
+            onError -> fail("Upload failed for a different reason.")
+        );
+
+        // Listen to Hub events to cancel when progress has been made
+        SubscriptionToken progressToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
+            if ("uploadProgress".equals(hubEvent.getName())) {
+                HubEvent<Float> progressEvent = (HubEvent<Float>) hubEvent;
+                Float progress = progressEvent.getData();
+                if (progress != null && progress > 0) {
+                    op.cancel();
+                }
+            }
+        });
+        subscriptions.add(progressToken);
+
+        // Listen to Hub events for cancel
+        SubscriptionToken cancelToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
```

Comment:
You want to be sure to setup all of your subscriptions _before_ doing the Storage operation. Right now, you have a race condition. One pony starts running, executing the code to do the S3 upload. And another pony starts running, forming these Hub subscriptions. It might be that the first pony dispatches the Hub event, before the second pony is ready to receive it. But, if you setup a subscription first, it is guaranteed to get whatever event(s) it is waiting for. (If they are sent ever, that is.)

Reply:
I would have liked to also, but I need the reference to operation object from storage upload so that I can call methods on it.

Reply:
Ah crap, yea I see what you mean. I guess you could do:

```
final AtomicReference<Cancelable> cancelableContainer = new AtomicReference<>();
Amplify.Hub.subscribe(channel, () -> {
   ....
   Cancelable cancelable = cancelableContainer.get();
   if (cancelable != null) cancelable.cancel();
   ...
}

cancelable.set(Amplify.Storage.doStuff(...));
```